vslib/src/sai_vs_hostintf.cpp

```diff
@@ -29,7 +67,414 @@ sai_status_t vs_send_hostif_packet(
+
+        if (size < 0)
+        {
+            SWSS_LOG_ERROR("failed to read from socket %d", info->packet_socket);
+            break;
+        }
+
+        // TODO examine packet for mac and possible generate fdb_event
+
+        if (write(info->tapfd, buffer, size) < 0)
+        {
+            SWSS_LOG_ERROR("failed to write to tap device %d", info->tapfd);
+            break;
+        }
+    }
+}
+
+void tap2veth_fun(std::shared_ptr<hostif_info_t> info)
+{
+    SWSS_LOG_ENTER();
+
+    unsigned char buffer[0x4000];
+
+    while (info->run_thread)
+    {
+        // TODO convert to non blocking using select
+
+        ssize_t size = read(info->tapfd, buffer, sizeof(buffer));
+
+        if (size < 0)
+        {
+            SWSS_LOG_ERROR("failed to read from tapfd: %d", info->tapfd);
+
+            break;
+        }
+
+        if (write(info->packet_socket, buffer, (int)size) < 0)
+        {
+            SWSS_LOG_ERROR("failed to write to socker %d", info->packet_socket);
+            break;
+        }
+    }
+}
+
+bool hostif_create_tap_veth_forwarding(
+        _In_ const std::string &tapname,
+        _In_ int tapfd)
+{
+    SWSS_LOG_ENTER();
+
+    // we assume here that veth devices were added by user before creating this
+    // host interface, vEthernetX will be used for packet transfer between ip
+    // namespaces
+
+    std::string vethname = "v" + tapname;
+
+    int packet_socket = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
+
+    if (packet_socket < 0)
+    {
+        SWSS_LOG_ERROR("failed to open packet socket, errno: %d", errno);
+
+        return false;
+    }
+
+    // bind to device
+
+    struct sockaddr_ll sock_address;
+
+    memset(&sock_address, 0, sizeof(sock_address));
+
+    sock_address.sll_family = PF_PACKET;
+    sock_address.sll_protocol = htons(ETH_P_ALL);
+    sock_address.sll_ifindex = if_nametoindex(vethname.c_str());
+
+    if (sock_address.sll_ifindex == 0)
+    {
+        SWSS_LOG_ERROR("failed to get interface index for %s", vethname.c_str());
+
+        close(packet_socket);
+
+        return false;
+    }
+
+    SWSS_LOG_INFO("interface index = %d %s\n", sock_address.sll_ifindex, vethname.c_str());
+
+    if (bind(packet_socket, (struct sockaddr*) &sock_address, sizeof(sock_address)) < 0)
+    {
+        SWSS_LOG_ERROR("bind failed on %s", vethname.c_str());
+
+        close(packet_socket);
+
+        return false;
+    }
+
+    std::shared_ptr<hostif_info_t> info = std::make_shared<hostif_info_t>();
+
+    hostif_info_map[tapname] = info;
+
+    info->packet_socket = packet_socket;
+    info->tapfd         = tapfd;
+    info->run_thread    = true;
+    info->e2t           = std::make_shared<std::thread>(veth2tap_fun, info);
+    info->t2e           = std::make_shared<std::thread>(tap2veth_fun, info);
+
+    info->e2t->detach();
+    info->t2e->detach();
+
+    SWSS_LOG_NOTICE("setup forward rule for %s succeeded", tapname.c_str());
+
+    return true;
+}
+
+sai_status_t vs_create_hostif_int(
```

Comment:
>vs_create_hostif_int [](start = 13, length = 20)

What is the use case for this function? Do you have a way to decently clean up? ie. stop the threads. I just feel the tricks to keep shared_pointer live is too smart...

Reply:
threads should be stopped in remove function, this is not implemented yet, see XXX/TODO comment that you commented